File: dotnet/src/SimpleOrm/MappingOptions.cs
namespace SimpleOrm;

/// <summary>
/// Configuration for metadata loading: the naming convention (default snake_case)
/// and explicit <see cref="EntityMapBuilder{T}"/> registrations, which take
/// precedence over attributes (§7.2: explicit → attribute → convention).
/// </summary>
public sealed class MappingOptions
{
    /// <summary>Shared default options: snake_case, no explicit registrations.</summary>
    public static MappingOptions Default { get; } = new();

    internal Dictionary<Type, Func<INamingConvention, EntityMap>> ExplicitMaps { get; } = [];

    public INamingConvention NamingConvention { get; set; } = SnakeCaseNamingConvention.Instance;

    /// <summary>Registers a manual map for <typeparamref name="T"/>, overriding its attributes and conventions.</summary>
    public MappingOptions Register<T>(EntityMapBuilder<T> builder)
        where T : class
    {
        ExplicitMaps[typeof(T)] = builder.Build;
        return this;
    }
}

File: php/src/Metadata/MappingOptions.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Metadata;

use SimpleOrm\Naming\NamingConvention;
use SimpleOrm\Naming\SnakeCaseNamingConvention;

/**
 * Configuration for metadata loading (§7.2): the naming convention (default
 * snake_case) and explicit registrations, which take precedence over
 * attributes and conventions (explicit → attribute → convention). Unlike the
 * C# reference's mutable `MappingOptions.Register<T>` (a deferred
 * `EntityMapBuilder<T>.Build` call), `$explicitMaps` holds already-built
 * {@see EntityMap} instances — PHP's `EntityMapBuilder::build()` takes no
 * convention argument, so building eagerly needs no deferral.
 */
final readonly class MappingOptions
{
    /** @param array<class-string, EntityMap> $explicitMaps */
    public function __construct(
        public NamingConvention $naming = new SnakeCaseNamingConvention(),
        public array $explicitMaps = [],
    ) {
    }

    /** Shared default options: snake_case, no explicit registrations. */
    public static function default(): self
    {
        return new self();
    }
}
